utilities/transactions/pessimistic_transaction.cc

```diff
@@ -1203,9 +1207,19 @@ Status PessimisticTransaction::ValidateSnapshot(
     PutFixed64(&ts_buf, read_timestamp_);
   }
 
-  return TransactionUtil::CheckKeyForConflicts(
+  bool found_record_for_key = false;
+  Status s = TransactionUtil::CheckKeyForConflicts(
       db_impl_, cfh, key.ToString(), snap_seq, ts_sz == 0 ? nullptr : &ts_buf,
-      false /* cache_only */);
+      /*cache_only=*/false, &found_record_for_key, /*snap_checker=*/nullptr,
+      /*min_uncommitted=*/kMaxSequenceNumber);
+  // If a record is found or the checking is not successful due to other
+  // failures, it's considered the key may exist in the db.
+  if (!keys_may_exist_in_db_ &&
```

Comment:
Is `ValidateSnapshot()` guaranteed to find records for a key that are older than the snapshot `snap_seq`? I thought it stops looking once it reaches a component that contains data older than `snap_seq`.

Reply:
That's a good question, yes you are right. This only reads back to the snapshot. So the timestamp validation in the transaction layer is only attempting to guarantee the user-defined timestamp invariant for the key across concurrently running transactions, not the whole DB.

MyRocks decided they need to disable this timestamp validation all together for the whole DB, because of difficulty of getting a proper read timestamp. So I will prioritize that and let me punt on this one and think more about it. Thanks for calling this out!